### DISCUSSION TITLE
Tree-shaking with Vite.js

### DISCUSSION THREAD

**Discussion by dylanalizon** [quasarframework/quasar#11142] [General - Components / Directives / etc] [answered]
Hey,

We are trying to use Quasar in a Vue 3 (Vite) project. So we use the new "Vite plugin" to install it. 

We noticed that after the import in the main.js file, the vendor.js file generated by Vite.js had grown a lot and contained all the basic components of the framework.

In the documentation, we can see that tree-shaking is handled automatically. 

We tried this: 

```js
myApp.use(Quasar, {
  components: {QInput} 
})
```

But it doesn't change anything.

So how can we apply the tree-shaking to avoid having components that are not used in the build ?

**Comment by rstoenescu** [comment, marked as answer]
You're making a lot of assumptions when writing this. What does "all the basic components" mean?
Tree-shaking is tested and guaranteed to work with our Vite plugin. Please post a repo so I can check out your usage and see if something is indeed wrong with the plugin (which I doubt, honestly).

**Reply by dylanalizon** [reply]
Hi,

we agree that the term "tree-shaking" means that in the built files there is only the code of the components used in our project?

I made a Github repository with a Vue 3 project, created with Vite and installed Quasar as indicated in the documentation.

If you open the dist/assets/vendor.ce6f4364.js file you can see that it contains some (or maybe all) of the components. For example if I search for "QBadge" or "QDialog" I can find the whole components even though I have not used them.

The Github repo: https://github.com/dylanalizon/quasar-vite
The vendor file: https://github.com/dylanalizon/quasar-vite/blob/main/dist/assets/vendor.ce6f4364.js

As a solution, to ensure that the build does not get too big, we only use the Vite plugin without initializing Quasar in the main.ts file, which results in importing only the components used at the time of the build.

However, we are forced to create a global property within Vue to define the dark mode for example.

The property look like this:

```javascript
app.config.globalProperties.$q = {
  dark: {
    isActive: false,
  },
};
```

Is this the right way to do it? Or is there a better way?

Please know that I am not questioning your work. I find this framework very complete and very well documented.

**Reply by rstoenescu** [reply]
Thanks for your reply.
The key point here is Typescript. Forgot to enable tree-shaking for it also.

Fix will be available shortly in "@quasar/vite-plugin" v1.0.2.
